Frameworks/UIKit/_UIGestureCoordinator.mm

```diff
@@ -0,0 +1,457 @@
+    // we will rebuild DManip gesture list
+    NSMutableArray* _dManipRecognizers;
+
+    // gesture list used for booking up gestures to be removed in one Touch
+    // processing cycle. When processing a Touch event, some gestures can be
+    // cancelled or be recognized, thsu become ended. These gestures will be
+    // stored in this variable and then clean up. So when processing next touch
+    // we don't have to enumerate through ended gesture, improve enumeration efficience
+    NSMutableSet* _gesturesToRemove;
+
+    // gesture list used for booking gestures in possible state.  when one candidate
+    // gesture in _regularRecognizers becomes active we need enumerate this list to
+    // make sure that if all gestures in this list can be fired concurrently, if not,
+    // they will be transit to cancelled state and removed from this list.
+    // it improves gesture processing efficiency
+    NSMutableSet* _gesturesInPossibleState;
+}
+
++ (instancetype)singleton {
+    if (!g_UIGestureCoordinator) {
+        g_UIGestureCoordinator = [_UIGestureCoordinator new];
+    }
+
+    return g_UIGestureCoordinator;
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        _resetAllTrackingGestures = YES;
+        _gesturesToRemove = [NSMutableSet new];
+        _gesturesInPossibleState = [NSMutableSet new];
+        _regularRecognizers = [NSMutableArray new];
+        _dManipRecognizers = [NSMutableArray new];
+    }
+
+    return self;
+}
+
+// test override to provide custom gestures
+- (void)buildTrackingGesturesListWith:(NSMutableArray*)regularRecognizers
+                     PossibleGestures:(NSMutableSet*)gesturesInPossibleState
+                        DManipGesture:(NSMutableArray*)dManipRecognizers {
+    // by default do nothing
+}
+
+// build tracking gesture list given a view, if view is nil, it invokes buildTrackingGesturesListWith override to get
+// custom build list of gesture for testing
+- (void)_buildTrackingGesturesListForView:(UIView*)view {
+    if (view == nil) {
+        if (_resetAllTrackingGestures) {
+            _resetAllTrackingGestures = FALSE;
+            return [self buildTrackingGesturesListWith:_regularRecognizers
+                                      PossibleGestures:_gesturesInPossibleState
+                                         DManipGesture:_dManipRecognizers];
+        }
+    }
+
+    // Otherwise, build up gesture tracking list based on view
+    if (_resetAllTrackingGestures) {
+        _resetAllTrackingGestures = FALSE;
+
+        // Find gesture recognizers in the hierarchy, back-first
+        UIView* curView = view;
+        while (curView != nil) {
+            // Adding enabled gestures on each view into cooresponding tracking gesture list
+            for (UIGestureRecognizer* curgesture in curView->priv->gestures.get()) {
+                if ([curgesture isEnabled]) {
+                    // Separating gestures into regular (Non-DManip) and DManip Gesture
```

Comment:
fortunately no. _resetAllTrackingGestures is a switch basically guard against rebuid. if _resetAllTrackingGestures is set to TRUE, normally gets set when all finger lifted off screen.  Otherwise, once it sets to FALSE, stay as FALSE until reset.

---
In reply to: [100589248](https://github.com/Microsoft/WinObjC/pull/1909#discussion_r100589248) [](ancestors = 100589248)